File: 2023/day10/day10.py
#!/usr/bin/env python3
# [Day 10: Pipe Maze](https://adventofcode.com/2023/day/10)

import argparse
from collections import deque
from pathlib import Path

from shapely.geometry import Point
from shapely.geometry.polygon import Polygon

parser = argparse.ArgumentParser()
parser.add_argument("--elapsed", action="store_true")
parser.add_argument("--png", action="store_true")
parser.add_argument("-o", "--output", type=Path, default="pipemaze.png")
parser.add_argument("-s", "--scale", type=int, default=5)
parser.add_argument("-v", "--verbose", action="store_true")
parser.add_argument("filename", type=Path, nargs="?", default="input.txt")
args = parser.parse_args()
data = args.filename.read_text().strip()
lines = data.splitlines()

_grid = [[c for c in line] for line in lines]

sx = len(_grid[0])
sy = len(_grid)


def grid(x, y):
    if 0 <= x < sx and 0 <= y < sy:
        return _grid[y][x]
    return "."


for y in range(sy):
    for x in range(sx):
        if grid(x, y) == "S":
            starting = (x, y)


visited = set()
q = deque()

q.append(starting)
pt = []
while q:
    x, y = q.pop()

    if (x, y) in visited:
        continue
    visited.add((x, y))

    pt.append(Point(x, y))

    c = grid(x, y)

    if grid(x, y + 1) in "|LJ" and c in "|7FS":
        q.append((x, y + 1))

    if grid(x, y - 1) in "|7F" and c in "|LJS":
        q.append((x, y - 1))

    if grid(x - 1, y) in "-FL" and c in "-J7S":
        q.append((x - 1, y))

    if grid(x + 1, y) in "-7J" and c in "-FLS":
        q.append((x + 1, y))

part1 = len(visited) // 2


interior = set()
po = Polygon(pt)
part2 = 0
for y in range(sy):
    for x in range(sx):
        p = (x, y)
        if p in visited:
            continue
        if po.contains(Point(*p)):
            part2 += 1
            interior.add((x, y))


def prt():
    for y in range(sy):
        s = ""
        for x in range(sx):
            if (x, y) in interior:
                c = "\033[93mI\033[0m"
            elif (x, y) not in visited:
                c = " "
            else:
                c = grid(x, y)

                if c == "S":
                    c = "\033[32mS\033[0m"
                elif c == ".":
                    c = " "
                else:
                    if c == "F":
                        c = "┌"
                    if c == "J":
                        c = "┘"
                    if c == "L":
                        c = "└"
                    if c == "7":
                        c = "┐"
                    if c == "-":
                        c = "─"
                    if c == "|":
                        c = "│"
                    c = f"\033[95m{c}\033[0m"
            s += c
        print(s)


def png():
    from PIL import Image

    S = args.scale

    im = Image.new("RGB", (sx * S, sy * S))

    def square(x, y, color):
        for k in range((S - 2) * (S - 2)):
            im.putpixel((x * S + 1 + k % (S - 2), y * S + 1 + k // (S - 2)), color)

        # for k in range(S * S):
        #     im.putpixel((x * S + k % S, y * S + k // S), color)

    def line(x, y, d):
        im.putpixel((x * S + S // 2, y * S + S // 2), (255, 0, 255))
        for c in d:
            for k in range(S // 2):
                match c:
                    case ">":
                        im.putpixel((x * S + S - 1 - k, y * S + S // 2), (255, 0, 255))
                    case "<":
                        im.putpixel((x * S + k, y * S + S // 2), (255, 0, 255))
                    case "v":
                        im.putpixel((x * S + S // 2, y * S + S - 1 - k), (255, 0, 255))
                    case "^":
                        im.putpixel((x * S + S // 2, y * S + k), (255, 0, 255))

    for y in range(sy):
        for x in range(sx):
            if (x, y) in interior:
                square(x, y, (255, 255, 0))

            elif (x, y) not in visited:
                pass

            else:
                c = grid(x, y)

                if c == "S":
                    # background square to mark the starting position
                    square(x, y, (40, 130, 40))

                    # guess the missing pipe shape
                    d = ""
                    if grid(x, y - 1) in "F7|":
                        d += "^"
                    if grid(x, y + 1) in "LJ|":
                        d += "v"
                    if grid(x - 1, y) in "FL-":
                        d += "<"
                    if grid(x + 1, y) in "7J-":
                        d += ">"
                    line(x, y, d)

                elif c == ".":
                    pass

                else:
                    d = {"F": "v>", "J": "<^", "L": ">^", "7": "<v", "-": "<>", "|": "^v"}[c]
                    line(x, y, d)

    im.save(args.output)


if args.png:
    png()

if args.verbose:
    prt()

print(part1)
print(part2)

File: 2023/day10/day10.rs
//! [Day 10: Pipe Maze](https://adventofcode.com/2023/day/10)

use geo::Contains;
use geo::Polygon;
use geo_types::{Coord, LineString};

use std::collections::{HashSet, VecDeque};

use itertools::iproduct;

struct Puzzle {
    grid_data: Vec<Vec<char>>,
    sx: i32,
    sy: i32,
    start_position: (i32, i32),
    points: Vec<(i32, i32)>,
}

impl Puzzle {
    const fn new() -> Self {
        Self {
            grid_data: vec![],
            sx: 0,
            sy: 0,
            start_position: (0, 0),
            points: vec![],
        }
    }

    /// Get the puzzle input.
    fn configure(&mut self, path: &str) {
        let data = std::fs::read_to_string(path).unwrap();

        self.points.clear();
        self.grid_data.clear();
        self.sx = 0;
        for line in data.lines() {
            if self.sx == 0 {
                self.sx = i32::try_from(line.len()).unwrap();
            } else {
                assert_eq!(self.sx, i32::try_from(line.len()).unwrap());
            }

            self.grid_data.push(line.chars().collect());
        }
        self.sy = i32::try_from(self.grid_data.len()).unwrap();

        for (x, y) in iproduct!(0..self.sx, 0..self.sy) {
            if self.grid(x, y) == 'S' {
                self.start_position = (x, y);
                break;
            }
        }

        self.maze();
    }

    fn grid(&self, x: i32, y: i32) -> char {
        if 0 <= x && x < self.sx && 0 <= y && y < self.sy {
            let x = usize::try_from(x).unwrap();
            let y = usize::try_from(y).unwrap();
            self.grid_data[y][x]
        } else {
            '.'
        }
    }

    fn maze(&mut self) {
        let mut visited = HashSet::new();
        let mut queue = VecDeque::new();

        queue.push_back(self.start_position);

        while !queue.is_empty() {
            let p = queue.pop_back().unwrap();

            if visited.contains(&p) {
                continue;
            }
            visited.insert(p);

            self.points.push(p);

            let (x, y) = p;
            let c = self.grid(x, y);

            if "|LJ".contains(self.grid(x, y + 1)) && "|7FS".contains(c) {
                queue.push_back((x, y + 1));
            }

            if "|7F".contains(self.grid(x, y - 1)) && "|LJS".contains(c) {
                queue.push_back((x, y - 1));
            }

            if "-FL".contains(self.grid(x - 1, y)) && "-J7S".contains(c) {
                queue.push_back((x - 1, y));
            }

            if "-7J".contains(self.grid(x + 1, y)) && "-FLS".contains(c) {
                queue.push_back((x + 1, y));
            }
        }
    }

    /// Solve part one.
    fn part1(&self) -> usize {
        self.points.len() / 2
    }

    /// Solve part two.
    fn part2(&self) -> u32 {
        let line_string = self
            .points
            .iter()
            .map(|&(x, y)| Coord { x, y })
            .collect::<LineString<i32>>();

        let polygon = Polygon::new(line_string, vec![]);

        let mut n = 0;
        for (x, y) in iproduct!(0..self.sx, 0..self.sy) {
            let p = Coord { x, y };
            if polygon.contains(&p) {
                n += 1;
            }
        }

        n
    }
}

fn main() {
    let args = aoc::parse_args();
    let mut puzzle = Puzzle::new();
    puzzle.configure(args.path.as_str());
    println!("{}", puzzle.part1());
    println!("{}", puzzle.part2());
}

/// Test from puzzle input
#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test01_1() {
        let mut puzzle = Puzzle::new();
        puzzle.configure("test1.txt");
        assert_eq!(puzzle.part1(), 4);
    }
    #[test]
    fn test01_2() {
        let mut puzzle = Puzzle::new();
        puzzle.configure("test2.txt");
        assert_eq!(puzzle.part1(), 8);
    }

    #[test]
    fn test02_3() {
        let mut puzzle = Puzzle::new();
        puzzle.configure("test3.txt");
        assert_eq!(puzzle.part2(), 4);
    }

    #[test]
    fn test02_4() {
        let mut puzzle = Puzzle::new();
        puzzle.configure("test4.txt");
        assert_eq!(puzzle.part2(), 4);
    }
    #[test]
    fn test02_5() {
        let mut puzzle = Puzzle::new();
        puzzle.configure("test5.txt");
        assert_eq!(puzzle.part2(), 8);
    }

    #[test]
    fn test02_6() {
        let mut puzzle = Puzzle::new();
        puzzle.configure("test6.txt");
        assert_eq!(puzzle.part2(), 10);
    }
}
